### PR TITLE
[GOBBLIN-1233] Add case-aware support in WhitelistBlacklist and other small fixes

### PR DESCRIPTION
**Pull request by zxcware**
… small fixes

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. 
    - https://issues.apache.org/jira/browse/GOBBLIN-1233

### Description
- [x] Here are some details about my PR:
 - `WhitelistBlacklist` today is case agnostic. The PR adds an option to make it case sensitive
 - Fix Bug: [GOBBLIN-831](https://issues.apache.org/jira/browse/GOBBLIN-831). On zero work unit, `KafkaWorkUnitPacker` will trigger NPE
 - Fix Bug: `DatasetCleanerTask` doesn't set working state on task failure or succeess
 - Improvement: Allow `AsyncHttpWriter` to retry on any transmission exception besides IOException

### Tests
- [x] My PR adds the following unit tests:
 - `WhitelistBlacklistTest.testCaseAware()` covers case sensitive support. The other tests verifies that the change is backward compatible
 - `KafkaWorkUnitPackerTest.testPackEmptyWorkUnit()` verifies no NPE given 0 work units
 - `AsyncHttpWriterTest.testClientError()` verifies retries are done on RuntimeException other than IOException.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java [hunk 1]
```diff
@@ -53,23 +53,36 @@
 
   public static final String WHITELIST = "whitelist";
   public static final String BLACKLIST = "blacklist";
+  public static final String IGNORE_CASE = "whitelistBlacklist.ignoreCase";
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java:56]: Please update the javadoc explaining IGNORE_CASE

File: gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java [hunk 2]
```diff
@@ -53,23 +53,36 @@
...
+    this(whitelist, blacklist, true);
+  }
+
+  public WhitelistBlacklist(String whitelist, String blacklist, boolean ignoreCase) throws IOException {
     this.whitelistMultimap = HashMultimap.create();
     this.blacklistMultimap = HashMultimap.create();
 
-    populateMultimap(this.whitelistMultimap, whitelist.toLowerCase());
-    populateMultimap(this.blacklistMultimap, blacklist.toLowerCase());
+    this.ignoreCase = ignoreCase;
```
Comment (arjun4084346) [inline gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java:78]: Nit a line break could be after setting ignoreCase and not before it.

File: gobblin-core/src/main/java/org/apache/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java [hunk 3]
```diff
@@ -74,6 +76,7 @@
   private final String writerPartitionSuffix;
   private final DatePartitionType granularity;
   private final DateTimeZone timeZone;
+  @Getter
```
Comment (arjun4084346) [inline gobblin-core/src/main/java/org/apache/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java:79]: required?

Comment (zxcware) [reply gobblin-core/src/main/java/org/apache/gobblin/writer/partitioner/TimeBasedWriterPartitioner.java:79]: Yes. We have a row check policy that requires time unit info from partitioner to compare record time with current time.

File: gobblin-data-management/src/main/java/org/apache/gobblin/runtime/retention/DatasetCleanerTask.java [hunk 4]
```diff
@@ -47,8 +51,15 @@ public void run() {
...
+      this.workingState = WorkUnitState.WorkingState.SUCCESSFUL;
     } catch (IOException e) {
+      this.workingState = WorkUnitState.WorkingState.FAILED;
       throw new RuntimeException(e);
     }
   }
+
+  @Override
+  public void commit() {
+    log.info("task {} commits with state {}", this.taskContext.getTaskState().getTaskId(), this.workingState);
```
Comment (ZihanLi58) [inline gobblin-data-management/src/main/java/org/apache/gobblin/runtime/retention/DatasetCleanerTask.java:63]: Just curious, if this task runs in mapper or driver? I think most gobblin retention job runs on driver, if that's the case, what's the taskID here?

File: gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java (hunk 1 above)
Comment (zxcware) [reply gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java:62]: fixed

File: gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java (hunk 2 above)
Comment (zxcware) [reply gobblin-utility/src/main/java/org/apache/gobblin/data/management/copy/hive/WhitelistBlacklist.java:83]: Fixed

File: gobblin-modules/gobblin-http/src/main/java/org/apache/gobblin/writer/AsyncHttpWriter.java [hunk 5]
```diff
@@ -83,7 +83,7 @@ protected void dispatch(Queue<BufferedRecord<D>> buffer) throws DispatchExceptio
     while (attempt < maxAttempts) {
       try {
           response = httpClient.sendRequest(rawRequest);
-      } catch (IOException e) {
+      } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (zxcware) [reply gobblin-modules/gobblin-http/src/main/java/org/apache/gobblin/writer/AsyncHttpWriter.java:86]: To allow retry on any exception.

File: gobblin-data-management/src/main/java/org/apache/gobblin/runtime/retention/DatasetCleanerTask.java (hunk 4 above)
Comment (zxcware) [reply gobblin-data-management/src/main/java/org/apache/gobblin/runtime/retention/DatasetCleanerTask.java:63]: regardless where it runs, a task will have a task id, right?

**Pull request closed without merging**